cpmpy/transformations/linearize.py

```diff
@@ -255,71 +255,87 @@ def only_positive_bv(lst_of_expr):
+
+            # LHS is sum/wsum/var
+            comp_name = cpm_expr.name
             lhs, rhs = cpm_expr.args
-            if isinstance(lhs, Comparison) and cpm_expr.name == "==":  # reification of comparison
-                lhs = canonical_comparison(lhs)[0]
-            elif is_num(lhs) or isinstance(lhs, _NumVarImpl) or (isinstance(lhs, Operator) and lhs.name in {"sum", "wsum"}):
-                # bring all vars to lhs
-                lhs2 = []
-                if isinstance(rhs, _NumVarImpl):
-                    lhs2, rhs = [-1 * rhs], 0
-                elif isinstance(rhs, Operator) and rhs.name == "sum":
-                    lhs2, rhs = [-1 * b if isinstance(b, _NumVarImpl) else 1 * b.args[0] for b in rhs.args
-                                 if isinstance(b, _NumVarImpl) or isinstance(b, Operator)], \
-                                 sum(b for b in rhs.args if is_num(b))
-                elif isinstance(rhs, Operator) and rhs.name == "wsum":
-                    lhs2, rhs = [-a * b for a, b in zip(rhs.args[0], rhs.args[1])
-                                    if isinstance(b, _NumVarImpl)], \
-                                    sum(-a * b for a, b in zip(rhs.args[0], rhs.args[1])
-                                    if not isinstance(b, _NumVarImpl))
-                if isinstance(lhs, Operator) and lhs.name == "sum":
-                    lhs, rhs = sum([1 * a for a in lhs.args] + lhs2), rhs
-                elif isinstance(lhs, _NumVarImpl) or (isinstance(lhs, Operator) and lhs.name == "wsum"):
-                    lhs, rhs = lhs + lhs2, rhs
-                else:
-                    raise ValueError(
-                        f"unexpected expression on lhs of expression, should be sum,wsum or intvar but got {lhs}")
 
-                assert not is_num(lhs), "lhs cannot be an integer at this point!"
+            # put all vars to lhs
+            if not is_num(rhs) and (isinstance(lhs, _NumVarImpl) or lhs.name =="sum" or lhs.name == "wsum"):
 
-                # bring all const to rhs
-                if lhs.name == "sum":
-                    new_args = []
-                    for i, arg in enumerate(lhs.args):
-                        if is_num(arg):
-                            rhs -= arg
-                        else:
-                            new_args.append(arg)
-                    lhs = Operator("sum", new_args)
+                if isinstance(rhs, Operator):
+                    if rhs.name == "sum":
+                        extra_weights = [-1] * len(rhs.args)
+                        extra_vars = rhs.args
+                    elif rhs.name == "wsum":
+                        extra_weights=  [-w for w in rhs.args[0]]
+                        extra_vars = rhs.args[1]
+                elif isinstance(rhs, _NumVarImpl):
+                    extra_weights = [-1]
+                    extra_vars = [rhs]
 
+                if lhs.name == "sum":
+                    lhs_weights = [1]*len(lhs.args) + extra_weights
```

Comment:
extra_weights will not be defined if the rhs is an operator other than sum or wsum (like div or mul or whatever)